### ISSUE TITLE
Feature request: sqlite_backup_to_memory

### ISSUE THREAD

**Issue by hoehrmann** [DBD-SQLite/DBD-SQLite#30]
If SQLite allows creating a backup to the special `:memory:` filename, it would be nice if that was exposed through `DBD::SQLite` with, for instance, a `sqlite_backup_to_memory` method that returns a DBI handle to the in-memory database. At the moment, as I understand it, one would have to backup to a temp file, and then restore the backup to a `:memory:` database, which is a bit inefficient and leaves a lot of room for errors. If there is a better way that I've missed, a pointer in the documentation would be nice.

**Comment by charsbar** [comment]
@hoehrmann  Could you explain *why* it would be nice a little more with some use case(s)? Why don't you just use ```:memory:``` database from the start, and then back up to a file when necessary?

**Comment by hoehrmann** [comment]
@charsbar: One use case would be an automated test that checks if you have all the right TRIGGERs in the database schema. Create a temporary in-memory database, fill it with computed random data, and then run various DELETEs and SELECTs on a temporary in-memory copy of the test database.

My http://search.cpan.org/perldoc/Graph::Feather module uses SQLite as "internal data structure". I might want offer a cloning method without forcing users of the module to touch persistent storage. I could do that through the module's public interfaces, but it would be much faster and less error-prone to let the driver simply clone the database in memory.

It seems SQLite makes this fairly easy, but since `sqlite_backup_to_file` closes the new database rather than returning a handle to it, in-memory databases disappear, leaving only temporary files as workaround.

**Comment by hoehrmann** [comment]
Ah, this appears to work:

```perl
my $dbh1 = DBI->connect('dbi:SQLite:uri=file:memdb1?mode=memory&cache=shared');
$dbh1->{RaiseError} = 1;

local $dbh1->{sqlite_allow_multiple_statements} = 1;

$dbh1->do(q{
  CREATE TABLE x(y);
  INSERT INTO x(y) VALUES(3);
});

$dbh1->sqlite_backup_to_file('file:memdb2?mode=memory&cache=shared');

my $dbh2 = DBI->connect('dbi:SQLite:uri=file:memdb2?mode=memory&cache=shared');

$dbh2->do(q{
  ATTACH DATABASE 'file:memdb1?mode=memory&cache=shared' AS memdb1;
});

print Dump $dbh2->selectall_arrayref(q{
  SELECT * FROM memdb1.x
});
```

**Comment by charsbar** [comment]
@hoehrmann, could you try https://github.com/DBD-SQLite/DBD-SQLite/tree/backup_between_dbhs branch and see if backup_from_dbh/backup_to_dbh work for you? (You can find example in t/34_online_backup.t)

**Comment by charsbar** [comment]
Shipped 1.61_02 with the above feature anyway. Thanks.
